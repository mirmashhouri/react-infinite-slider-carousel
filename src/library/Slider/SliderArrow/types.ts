import type { FC } from 'react';

export type SliderComponentArrow = FC<{
  hasLeft: boolean;
  hasMoreSlidesOnRight: boolean;
  handleClick: () => void;
  isLeft?: boolean;
}>;
