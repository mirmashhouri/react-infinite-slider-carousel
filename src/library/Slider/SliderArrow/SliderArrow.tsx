import React, { memo } from 'react';

import { SliderComponentArrow } from './types';

export const SliderArrow: SliderComponentArrow = memo(({
  isLeft,
  hasLeft,
  hasMoreSlidesOnRight,
  handleClick,
}) => (
  <button
    className={ `slider__arrow ${
      isLeft ? 'slider__arrow-left' : 'slider__arrow-right'
    } ${
      isLeft && !hasLeft ? 'slider__hide' : ''
    } ${
      !isLeft && !hasMoreSlidesOnRight ? 'slider__hide' : ''
    }` }
    onClick={ handleClick }
  >
    {
      isLeft ?
        <span>⧀</span>
        : <span>⧁</span>
    }
  </button>
));
