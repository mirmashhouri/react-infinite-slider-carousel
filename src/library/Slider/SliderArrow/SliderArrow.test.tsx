/* eslint-disable import/no-extraneous-dependencies */
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SliderArrow } from './SliderArrow';

describe('SliderArrow component', () => {
  const leftArrow = 'slider__arrow-left';
  const leftArrowIcon = 'bi-chevron-left';
  const rightArrow = 'slider__arrow-right';
  const rightArrowIcon = 'bi-chevron-right';
  const hideArrow = 'slider__hide';
  const buttonElement = 'button';

  const mockCallBack = jest.fn();

  const Component = (hasLeft: boolean, hasMoreSlidesOnRight: boolean, isLeft?: boolean ) => (
    <SliderArrow
      hasLeft={ hasLeft }
      hasMoreSlidesOnRight={ hasMoreSlidesOnRight }
      isLeft={ isLeft }
      handleClick={ mockCallBack }
    />
  );

  it('Render left arrow', () => {
    const rendered = render(Component(true, false, true));

    //Expect to render left arrow
    expect(
      rendered.container.getElementsByClassName(leftArrow),
    ).toHaveLength(1);

    //Expect to render left arrow icon
    expect(
      rendered.container.getElementsByClassName(leftArrowIcon),
    ).toHaveLength(1);
  });

  it('Render right arrow', () => {
    const rendered = render(Component(false, true));

    //Expect to render right arrow
    expect(
      rendered.container.getElementsByClassName(rightArrow),
    ).toHaveLength(1);

    //Expect to render right arrow icon
    expect(
      rendered.container.getElementsByClassName(rightArrowIcon),
    ).toHaveLength(1);
  });

  it('Test left arrow hide', () => {
    //Expect to hide left arrow
    expect(
      render(Component(false, true, true)).container.getElementsByClassName(hideArrow),
    ).toHaveLength(1);
  });

  it('Test right arrow hide', () => {
    //Expect to hide right arrow
    expect(
      render(Component(false, false, false)).container.getElementsByClassName(hideArrow),
    ).toHaveLength(1);
  });

  it('Test button and handleClick', () => {
    render(Component(false, true, true));

    userEvent.click(screen.getByRole(buttonElement));

    //Expect handleClick to work properly
    expect(mockCallBack).toBeCalledTimes(1);
  });

  it('Render to match snapshot', () => {
    //Expect to render left Arrow
    expect(
      render(Component(true, false, true)).container,
    ).toMatchSnapshot();

    //Expect to render rightArrow
    expect(
      render(Component(true, false, false)).container,
    ).toMatchSnapshot();
  });
});