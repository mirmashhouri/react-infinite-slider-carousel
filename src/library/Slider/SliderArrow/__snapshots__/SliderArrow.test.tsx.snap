// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SliderArrow component Render to match snapshot 1`] = `
<div>
  <button
    class="slider__arrow slider__arrow-left  "
  >
    <i
      class="bi bi-chevron-left"
    />
  </button>
</div>
`;

exports[`SliderArrow component Render to match snapshot 2`] = `
<div>
  <button
    class="slider__arrow slider__arrow-right  slider__hide"
  >
    <i
      class="bi bi-chevron-right"
    />
  </button>
</div>
`;
